drafts/web-editor-monaco-with-language-server-support.md

```diff
@@ -0,0 +1,971 @@
+
+## Creating the Language Server
+Full Backend implementation can be found at the below GitHub repo
+
+[HelloLS Websocket Launcher](https://github.com/NipunaMarcus/hellols/tree/websocket-launcher)
+
+> **Note**: Here the LSP4J version is bumped to 0.21.0. If you want you can go higher. Previously we were using LSP4J 0.9.0.
+
+### Step 1: Language Server Implementation
+For the Language server implementation here I'm using [LSP4J](https://github.com/eclipse-lsp4j/lsp4j). As you can see there are three classes available under the language-server module. So these three classes are added to implement the three main interfaces provided by LSP4J as to the breakdown given in the LSP specification to support general cases, language features, and workspace management.
+
+* HelloLanguageServer.java — This class implements the interface available in the LSP4J called LanguageServer which contains the general functionality of the language server such as initializing the language server, shutting down the language server and so on … Also if the LS needs to publish the diagnostics(compilation error and semantic errors) back to the Client (in our case VSCode plugin) LS needs to be client aware. To make the LS client aware we need to implement the LanguageClientAware interface which allows LS to get the language client instance.
+* HelloTextDocumentService.java — This class implements the interface available in the LSP4J called TextDocumentService which contains the language features and the text synchronization endpoints explained in the LSP spec.
+* HelloWorkspaceService.java — This class implements the interface available in the LSP4J called WorkspaceService which contains the workspace features such as workspace symbol and configuration changes…
+
+Let's get to the implementation.
+
+If you look at the `pom.xml` of the project you can see that there is a dependency for `LSP4J`.
+
+```xml
+<dependency>
+    <groupId>org.eclipse.lsp4j</groupId>                  
+    <artifactId>org.eclipse.lsp4j</artifactId>
+    <version>${lsp4j.version}</version>
+</dependency>
+```
+
+So this is the library that we will use to implement our Language Server for Hello Language.
+
+So first let’s implement the `HelloLanguageServer.java` as it is the entry point to the LS.
+
+```java
+package langserver;
+
+import org.eclipse.lsp4j.CompletionOptions;
+import org.eclipse.lsp4j.InitializeParams;
+import org.eclipse.lsp4j.InitializeResult;
+import org.eclipse.lsp4j.ServerCapabilities;
+import org.eclipse.lsp4j.TextDocumentSyncKind;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4j.services.LanguageClientAware;
+import org.eclipse.lsp4j.services.LanguageServer;
+import org.eclipse.lsp4j.services.TextDocumentService;
+import org.eclipse.lsp4j.services.WorkspaceService;
+
+import java.util.concurrent.CompletableFuture;
+
+public class HelloLanguageServer implements LanguageServer, LanguageClientAware {
```

Comment:
I was a bit confused of the `Hello`-prefix of those classes - would it be worth naming them more after their intetions, e.g. `LanguageServer`, `LanguageFeaturesService` , `WorkspaceManagementService`